pkg/build/builder/docker.go

```diff
@@ -103,7 +106,9 @@ func (d *DockerBuilder) Build() error {
 			if err != docker.ErrNoSuchImage {
 				return err
 			}
-			imageExists = false
+			if !stageNames.Has(imageName) {
```

Comment:
i think you can do this check *before* doing the docker inspect.  there's no need to do the docker inspect if you're going to ignore the result (because the name is a stage name)

Reply:
also you need to think through the scenario of:

```
FROM missingimage as stage1
RUN foo
FROM stage1 as missingimage
RUN bar
```

your logic would "allow" that (it wouldn't attempt to pull `missingimage` in advance, even if forcepull is set), whereas it should not (it should attempt to pull it and then fail if it can't).

So there is an ordering implication.